base/quadgk.jl

```diff
@@ -92,32 +92,35 @@ function do_quadgk{Tw}(f, s, n, ::Type{Tw}, abstol, reltol, maxevals, nrm)
         s1 = s[1]; s2 = s[end]; inf1 = isinf(s1); inf2 = isinf(s2)
         if inf1 || inf2
             if inf1 && inf2 # x = t/(1-t^2) coordinate transformation
-                return do_quadgk(t -> begin t2 = t*t; den = 1 / (1 - t2);
+                return do_quadgk_no_inf(t -> begin t2 = t*t; den = 1 / (1 - t2);
                                             f(t*den) * (1+t2)*den*den; end,
                                  map(x -> isinf(x) ? copysign(one(x), x) : 2x / (1+hypot(1,2x)), s),
                                  n, Tw, abstol, reltol, maxevals, nrm)
             end
-            s0,si = inf1 ? (s2,s1) : (s1,s2)
+            s0::eltype(s),si::eltype(s) = inf1 ? (s2,s1) : (s1,s2)
             if si < 0 # x = s0 - t/(1-t)
-                return do_quadgk(t -> begin den = 1 / (1 - t);
+                return do_quadgk_no_inf(t -> begin den = 1 / (1 - t);
                                             f(s0 - t*den) * den*den; end,
                                  reverse!(map(x -> 1 / (1 + 1 / (s0 - x)), s)),
                                  n, Tw, abstol, reltol, maxevals, nrm)
             else # x = s0 + t/(1-t)
-                return do_quadgk(t -> begin den = 1 / (1 - t);
+                return do_quadgk_no_inf(t -> begin den = 1 / (1 - t);
                                             f(s0 + t*den) * den*den; end,
                                  map(x -> 1 / (1 + 1 / (x - s0)), s),
                                  n, Tw, abstol, reltol, maxevals, nrm)
             end
         end
     end
+    return do_quadgk_no_inf(f, s, n, Tw, abstol, reltol, maxevals, nrm)
+end
 
+function do_quadgk_no_inf{Tw}(f, s, n, ::Type{Tw}, abstol, reltol, maxevals, nrm)
     key = rulekey(Tw,n)
-    x,w,gw = haskey(rulecache, key) ? rulecache[key] :
+    x::Vector{Tw},w::Vector{Tw},gw::Vector{Tw} = haskey(rulecache, key) ? rulecache[key] :
        (rulecache[key] = kronrod(Tw, n))
-    segs = Segment[]
-    for i in 1:length(s) - 1
-        heappush!(segs, evalrule(f, s[i],s[i+1], x,w,gw, nrm), Reverse)
+    segs = [evalrule(f, s[1],s[2], x,w,gw, nrm)]
+    for i in 2:length(s) - 1
+        heappush!(segs, evalrule(f, s[i], s[i+1], x,w,gw, nrm), Reverse)
```

Comment:
I'm worried that this will fail if `f` is not type-stable.  e.g. if it returns an `Int` sometimes and a `Float64` other times.  We want that case to work (even if it is slower).